### ISSUE TITLE
Need a method to choose JS engine.

### ISSUE THREAD

**Issue by holmescn** [reactjs/React.NET#360]
I am trying to develop React.NET project under macOS. The VroomJS need to be compiled and I am working on it.

As I reading the source of React.NET, it already add Chakra as a candidate JS engine, but the `GetEngine` method just tried each factory and return the first that works. I have tried to install JavaScriptEngineSwitcher and add it to `IServiceCollection`, but the `ReactEnvironment` remain try the `VroomJS` engine. I don't find how to replace the factory in the `TinyIoC` container.

I thought we can have a option for the chosen of JS engine. That's more convenient.

I am working on building my VroomJS engine on macOS. If that works, I could try to fix this.

**Comment by holmescn** [comment]
Perhaps a simple solution is to add a try-catch block to wrap the AddVroom method. then this factory will not be added since it failed to initialize VroomJS.

I found that the JavaScript Engine Switcher already update to v2.1.2 version, but the dependencies still be 2.0.0 of React.NET 3.0.0.

**Comment by holmescn** [comment]
A PR has been made.

**Comment by dima117** [comment]
I suggest to remove the engine selection functionality from the package and assigning responsibility for it to the user who configures React.NET. For example, it will be similarly to the selection of the database in the Entity Framework Core.

**Comment by Daniel15** [comment]
> I suggest to remove the engine selection functionality from the package and assigning responsibility for it to the user who configures React.NET

This is already handled - ReactJS.NET will only register engines if you have not registered them yourself: https://github.com/reactjs/React.NET/blob/master/src/React.Core/JavaScriptEngineFactory.cs#L360-L364. You just need to ensure you register the engines before registering ReactJS.NET.

**Comment by dima117** [comment]
I tried the following code (according to [this](https://github.com/reactjs/React.NET/issues/338#issuecomment-275593401) example):

```csharp
JsEngineSwitcher engineSwitcher = JsEngineSwitcher.Instance;
engineSwitcher.EngineFactories.Add(new JintJsEngineFactory());
engineSwitcher.DefaultEngineName = "JintJsEngine";

Initializer.Initialize(registration => registration.AsSingleton());

var container = React.AssemblyRegistration.Container;
container.Register<ICache, NullCache>();
container.Register<IFileSystem, SimpleFileSystem>();

ReactSiteConfiguration.Configuration
  .SetReuseJavaScriptEngines(false)
  .AddScript("Sample.jsx");
```

And I had an error in [this](https://github.com/reactjs/React.NET/blob/master/src/React.Core/JavaScriptEngineFactory.cs#L294) line (there is [existing issue](https://github.com/reactjs/React.NET/issues/379)).

I think, the error is in my code. But it's hard for me to know exactly where it is, because the configuration API uses implicit logic, which is dependent on global variables (JsEngineSwitcher).

**Comment by Taritsyn** [comment]
> I suggest to remove the engine selection functionality from the package and assigning responsibility for it to the user who configures React.NET. For example, it will be similarly to the DBMS selection in the Entity Framework Core.

@Daniel15 @dima117 @holmescn Good idea, but I would go further: need to remove from the React.Core all `JavaScriptEngineSwitcher.*` dependencies except for the JavaScriptEngineSwitcher.Core. Thus, users will not need to download unnecessary dependencies. Anyway we need to install manually the packages with native assemblies for the JavaScriptEngineSwitcher.V8 and JavaScriptEngineSwitcher.ChakraCore. I use this approach in the [Bundle Transformer](https://bundletransformer.codeplex.com/) and it works great for many years. This approach adds only one additional step in configuring of library - selection and installation of JavaScript engine.

All problems, that prevented to do it have already been resolved in latest versions of the JavaScript Engine Switcher:

1. For a long time already when using the MSIE JavaScript Engine in different threads there are no errors.
2. MSIE JavaScript Engine and JavaScriptEngineSwitcher.ChakraCore are run in a separate threads with increased stack size.
3. Created a module, that supports VroomJs - [JavaScriptEngineSwitcher.Vroom](http://www.nuget.org/packages/JavaScriptEngineSwitcher.Vroom/)
